Mathlib/Combinatorics/SimpleGraph/Diameter.lean

```diff
@@ -0,0 +1,202 @@
+
+## Todo
+
+- Prove that `G.egirth ≤ 2 * G.ediam + 1` and `G.girth ≤ 2 * G.diam + 1` when the diameter is
+  non-zero.
+
+-/
+
+namespace SimpleGraph
+variable {α : Type*} {G G' : SimpleGraph α}
+
+section ediam
+
+/--
+The extended diameter is the greatest distance between any two vertices, with the value `⊤` in
+case the distances are not bounded above, or the graph is not connected.
+-/
+noncomputable def ediam (G : SimpleGraph α) : ℕ∞ :=
+  ⨆ u, ⨆ v, G.edist u v
+
+lemma ediam_def : G.ediam = ⨆ p : α × α, G.edist p.1 p.2 := by
+  rw [ediam, iSup_prod]
+
+lemma edist_le_ediam {u v : α} : G.edist u v ≤ G.ediam :=
+  le_iSup₂ (f := G.edist) u v
+
+lemma ediam_eq_top_iff : G.ediam = ⊤ ↔ ∀ b < ⊤, ∃ u v, b < G.edist u v := by
+  simp only [ediam, iSup_eq_top, lt_iSup_iff]
+
+lemma ediam_eq_zero_of_subsingleton [Subsingleton α] : G.ediam = 0 := by
+  rw [ediam_def, ENat.iSup_eq_zero]
+  simpa [edist_eq_zero_iff, Prod.forall] using subsingleton_iff.mp ‹_›
+
+lemma nontrivial_of_ediam_ne_zero (h : G.ediam ≠ 0) : Nontrivial α := by
+  contrapose! h
+  rw [not_nontrivial_iff_subsingleton] at h
+  exact ediam_eq_zero_of_subsingleton
+
+lemma ediam_ne_zero [Nontrivial α] : G.ediam ≠ 0 := by
+  obtain ⟨u, v, huv⟩ := exists_pair_ne ‹_›
+  contrapose! huv
+  simp only [ediam, nonpos_iff_eq_zero, ENat.iSup_eq_zero, edist_eq_zero_iff] at huv
+  exact huv u v
+
+lemma subsingleton_of_ediam_eq_zero (h : G.ediam = 0) : Subsingleton α := by
+  contrapose! h
+  apply not_subsingleton_iff_nontrivial.mp at h
+  exact ediam_ne_zero
+
+lemma ediam_ne_zero_iff_nontrivial :
+    G.ediam ≠ 0 ↔ Nontrivial α :=
+  ⟨nontrivial_of_ediam_ne_zero, fun _ ↦ ediam_ne_zero⟩
+
+@[simp]
+lemma ediam_eq_zero_iff_subsingleton :
+    G.ediam = 0 ↔ Subsingleton α :=
+  ⟨subsingleton_of_ediam_eq_zero, fun _ ↦ ediam_eq_zero_of_subsingleton⟩
+
+lemma ediam_eq_top_of_not_connected [Nonempty α] (h : ¬G.Connected) : G.ediam = ⊤ := by
+  rw [connected_iff_exists_forall_reachable] at h
+  push_neg at h
+  obtain ⟨_, hw⟩ := h Classical.ofNonempty
+  rw [eq_top_iff, ← edist_eq_top_of_not_reachable hw]
+  exact edist_le_ediam
+
+lemma ediam_eq_top_of_not_preconnected (h : ¬G.Preconnected) : G.ediam = ⊤ := by
+  cases isEmpty_or_nonempty α
+  · exfalso
+    exact h <| IsEmpty.forall_iff.mpr trivial
+  · apply ediam_eq_top_of_not_connected
+    rw [connected_iff]
+    tauto
+
+lemma exists_edist_eq_ediam_of_ne_top [Nonempty α] (h : G.ediam ≠ ⊤) :
+    ∃ u v, G.edist u v = G.ediam :=
+  Prod.exists'.mp
+    <| ediam_def ▸ (ENat.sSup_mem_of_Nonempty_of_lt_top <| lt_top_iff_ne_top.mpr <| ediam_def ▸ h)
+
+lemma exists_edist_eq_ediam_of_finite [Nonempty α] [Finite α] :
+    ∃ u v, G.edist u v = G.ediam := by
+  let f : (α × α) → ℕ∞ := fun p ↦ G.edist p.1 p.2
+  by_cases h : G.ediam = ⊤
+  · have : ⨆ p, f p = ⊤ := by simp [f, ediam_def ▸ h]
+    have : ⊤ ∈ Set.range f := this ▸ (Set.range_nonempty _).csSup_mem (Set.finite_range _)
+    simp_all
+  · exact exists_edist_eq_ediam_of_ne_top h
```

Comment:
These two are general lemma about `iSup` respectively on
* `ENat`
* a `CompleteLinearOrder`

Can you make their proofs are one-liners from the respective APIs?

Reply:
I did the first one, I'm having trouble with the second one. Is the API of `CompleteLinearOrder` located in `Order/CompleteLattice`? If yes, then that file doesn't import `Finite`. Should I import it there?

Reply:
Yes, should be fine

Reply:
I've wrote the lemma in `ConditionallyCompleteLinearOrder` instead because it has the needed lemma `Nonempty.csSup_mem`, and because it is more general. However, I can't provide `iSup₂` version there because there is no `iSup_prod` lemmas. Is this fine for now? Another solution would be to move these lemmas to `CompleteLinearOrder`, but then work will need to be done to bring the equivalent of `Nonempty.csSup_mem` to there.

Reply:
Yes indeed I think it's fie